### PR TITLE
Peer Review

### PR DESCRIPTION
**Pull request by BolajiOlajide**
# Get Peer Review for the Dojo Space Allocator

## What does this PR do?
- #### It serves as my first output in checkpoint 1

## Description of Task to be completed?

-   #### I have a working prototype of Dojo Space Allocator

## How should this be manually tested?

-  #### Instructions have been inlcuded in my README file.

### CODE REVIEW

File: tests/test_fellow.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+class TestFellow(unittest.TestCase):
+	"""Test cases for the Fellow class"""
+
+	def setUp(self):
+		"""Setup a new Fellow object to be used for testing"""
+		self.fellow = Fellow('Brian','Mosigisi')
+
+	def test_fellow_instance(self):
+		"""Tests if the object created is an instance of the Fellow class"""
+		self.assertIsInstance(self.fellow,Fellow)
```
Comment (andela-mkamau) [inline tests/test_fellow.py]: I see you are testing if `self.fellow` is an instance of `Fellow` class. Is this really necessary?

File: tests/test_fellow.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		self.assertEqual(self.fellow.livingspace_allocated,False)
+
+	def test_person_livingspace(self):
+		"""Tests if the Fellow created has no Living Space"""
+		self.assertEqual(self.fellow.livingspace,'')
+
+	def test_check_fellow_id(self):
+		"""Check if the id assigned to a fellow is unique to only fellows"""
+		first_char = self.fellow.fellow_id[0]
+		self.assertEqual(first_char,'F')
```
Comment (andela-mkamau) [inline tests/test_fellow.py]: Am curious as to why you are only checking if the first character of the `id` is an `F`. Why not create a couple of Fellow objects and check whether their ids are not identical? What do you think about that?

File: tests/test_fellow.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+if __name__ == '__main__':
+	unittest.main()
```
Comment (andela-mkamau) [inline tests/test_fellow.py]: I like that you have kept your test simple and with singular concerns.

Could think of more tests for this Fellow class? More edge cases?

File: tests/test_livingspace.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+class TestLivingSpace(unittest.TestCase):
+	"""Test cases for the Living Space class"""
+
+	def setUp(self):
+		"""Setup a new Living Space object to be used for testing"""
+		self.livingspace = LivingSpace('turquoise')
+
+	def test_room_instance(self):
+		"""Tests if the object created is an instance of the LivingSpace class"""
+		self.assertIsInstance(self.livingspace,LivingSpace)
```
Comment (andela-mkamau) [inline tests/test_livingspace.py]: Again, I am of the opinion that this test is unnecessary.

File: tests/test_livingspace.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		"""Tests if the Living Space name is equal to 'turquoise' """
+		self.assertEqual(self.livingspace.name,'turquoise')
+
+	def test_room_type_variable(self):
+		"""Tests if the room type assigned to the room is 'LIVINGSPACE' """
+		self.assertEqual(self.livingspace.room_type,'LIVINGSPACE')
+
+	def test_room_max_occupant(self):
+		"""Tests if the maximum number of occupants in a Living Space is 6"""
+		self.assertEqual(self.livingspace.max_no_occupants,4)
```
Comment (andela-mkamau) [inline tests/test_livingspace.py]: This method's docstring says that the maximum number of occupants allowed is 6 yet you are testing that the maximum number set is 4. Please review that.

File: tests/test_livingspace.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+
+if __name__ == '__main__':
+	unittest.main()
```
Comment (andela-mkamau) [inline tests/test_livingspace.py]: Think of more tests for this class including edge cases.

File: tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	def test_office_instance(self):
+		"""Tests if the Office instance was correctly created"""
+		self.assertIsInstance(self.office,Office)
+
+	def test_livingspace_instance(self):
+		"""Tests if the LivingSpace instance was correctly created"""
+		self.assertIsInstance(self.livingspace,LivingSpace)
+
+if __name__ == '__main__':
+	unittest.main()
```
Comment (andela-mkamau) [inline tests/test_models.py]: Again, I think these tests are not really needed.

File: tests/test_staff.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+		self.assertEqual(self.staff.livingspace_allocated,False)
+
+	def test_person_livingspace(self):
+		"""Tests if the Staff created has no Living Space"""
+		self.assertEqual(self.staff.livingspace,'')
+
+	def test_check_staff_id(self):
+		"""Check if the id assigned to a fellow is unique to only Staff members"""
+		first_char = self.staff.staff_id[0]
+		self.assertEqual(first_char,'S')
```
Comment (andela-mkamau) [inline tests/test_staff.py]: Please review this method's  docstring. 
You could create a number of Staff objects and compare their `staff_id`s. I think that would make this test more conclusive.

**Review by andela-mkamau — COMMENTED** [review]
## General Comments
I am really impressed with the direction you are taking your project.

I have mostly looked at your tests. 
I see that you have began adding functionalities, especially within the Dojo class. In TDD fashion, you should ideally have written accompanying tests for those functionalities. 

- Update your .gitignore file to ignore pyc files
- Think of more tests cases. This is particularly important for the Dojo class where a lot i expected to take place.
- Try to make you commits smaller and more focussed. Have a look at [these recommendations](https://chris.beams.io/posts/git-commit/).
- Also checkout PEP8 guidelines

File: tests/test_fellow.py (hunk 1 above)
Comment (andela-bolajide) [inline tests/test_fellow.py]: I have corrected this.
You're right. Since i am already testing for the type there is no need to test the instance.

File: tests/test_fellow.py (hunk 2 above)
Comment (andela-bolajide) [inline tests/test_fellow.py]: Good point.
I've added a test to check if two different fellows have identical fellow id.

File: tests/test_livingspace.py (hunk 4 above)
Comment (andela-bolajide) [inline tests/test_livingspace.py]: I've corrected this.

File: tests/test_livingspace.py (hunk 5 above)
Comment (andela-bolajide) [inline tests/test_livingspace.py]: Corrected!

File: tests/test_models.py (hunk 7 above)
Comment (andela-bolajide) [inline tests/test_models.py]: Okay. I've removed them.

File: tests/test_staff.py (hunk 8 above)
Comment (andela-bolajide) [inline tests/test_staff.py]: Done.

File: models/dojo.py [hunk 9]
```diff
@@ -65,7 +80,7 @@ def add_person(self,arg):
             if self.full_name not in Dojo.staff:
                 new_staff = Staff(self.fname,self.lname)
                 Dojo.staff[self.full_name] = {}
```
Comment (hassan02) [inline models/dojo.py]: A better way to do this is to save the staff or fellow details in form of an object and not a dictionary. e.g `Dojo.staff[new_staff.full_name] = new_staff` or `Dojo.staff[new_staff.id] = new_staff`. With this, you wouldn't need lines 83-86 and line 101-104.

File: models/dojo.py [hunk 10]
```diff
@@ -99,72 +114,328 @@ def add_person(self,arg):
         else:
             print("Error! Individual must be either a fellow or a staff.")
 
-
-    def assign_room(self,arg):
+    def allocate_room(self,arg):
```
Comment (hassan02) [inline models/dojo.py]: You seem to be instantiating Dojo multiple times. That is bad practice. A better way to do this is to create a single instance with `dojo = Dojo()` and use `dojo` in place of `Dojo()`

**Review by hassan02 — CHANGES_REQUESTED** [review]

File: main.py [hunk 11]
```diff
@@ -15,7 +15,6 @@
 import sys
 import os
 import cmd
-from models.room import Room
 from docopt import docopt, DocoptExit
 from models.dojo import Dojo
```
Comment (mfalade) [inline main.py]: NIT: You should have a space between your inbuilt libraries, 3rd party libraries and project specific libraries import. 
Also, you should import in alphabetical order 

(that if you give a rat's ass about [PEP8 Standards](https://www.python.org/dev/peps/pep-0008/))

But seriously, this is recommended

File: main.py [hunk 12]
```diff
@@ -68,27 +67,50 @@ def do_create_room(self, arg):
     def do_add_person(self,arg):
         """Usage: add_person <person_fname> <person_lname> <FELLOW/STAFF> [<wants_accommodation>]"""
         check = Dojo().add_person(arg)
-        if check == True:
-                Dojo().assign_room(arg)
+        if check:
+                Dojo().allocate_room(arg)
```
Comment (mfalade) [inline main.py]: `Dojo().allocate_room` ???

Everytime you do `Dojo()`, you create a new instance of the class. 

You shouldn't do this.
It's bad design.

Create one instance instead 

```python
the_only_dojo_you_need = Dojo()
...
the_only_dojo_you_need.allocate_room(arg)
```

If you MUST access the instance method this way (i.e using the `Dojo().do_something format`), then you should implement the Dojo class using the singleton pattern 
(I challenge you to try this on the side just for fun 💪 )

File: main.py [hunk 13]
```diff
@@ -68,27 +67,50 @@ def do_create_room(self, arg):
...
         else:
-            pass
+            print("")
```
Comment (mfalade) [inline main.py]: Why print an empty string to the console ?

Comment (andela-bolajide) [inline main.py]: It's a method i was using to test. 
I've removed that empty and also the "else" condition as it serves no purpose here.

File: tests/test_office.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+	def test_room_is_empty(self):
+		"""Tests if the office initially has no occupant"""
+		self.assertEqual(self.office.occupant,[])
+
+	def test_room_number_of_occupants(self):
+		"""Tests if the office is empty"""
+		self.assertEqual(self.office.current_number,0)
+
+	def test_room_name(self):
```
Comment (andela-bolajide) [inline tests/test_office.py]: Corrected that. 
I will make sure i pay attention to these tiny details.

File: models/dojo.py [hunk 15]
```diff
@@ -16,7 +18,20 @@ class Dojo(object):
 
     def __init__(self):
```
Comment (andela-bolajide) [inline models/dojo.py]: I agree with you.
I've removed it.

File: models/dojo.py [hunk 16]
```diff
@@ -65,7 +80,7 @@ def add_person(self,arg):
             if self.full_name not in Dojo.staff:
```
Comment (andela-bolajide) [inline models/dojo.py]: I'm using the full name of the person as a key to store data in the dictionary.

File: models/dojo.py (hunk 10 above)
Comment (andela-bolajide) [inline models/dojo.py]: Awesome.
Done that.

File: models/dojo.py (hunk 9 above)
Comment (andela-bolajide) [inline models/dojo.py]: Done

File: main.py (hunk 11 above)
Comment (andela-bolajide) [inline main.py]: I've rearranged my import statements

File: models/fellow.py [hunk 17]
```diff
@@ -4,25 +4,29 @@
...
+		self.office=''
+		self.livingspace_allocated = False
+		self.livingspace = ''
 
-    def get_fellow_id(self):
-        """Returns unique id for Fellow by generating random string"""
-        id = 'F' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(7))
+	def get_fellow_id(self):
+		"""Returns unique id for Fellow by generating random string"""
+		id = 'F' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(7))
```
Comment (mfalade) [inline models/fellow.py]: `id` is a built in Python keyword and shouldn't be used as a variable name.

`fellow_id` is a more suitable name.

File: models/fellow.py [hunk 18]
```diff
@@ -4,25 +4,29 @@
...
-        """Checks if the random id created already exists in the array"""
-        while id in self.fellow_data:
-            """if the array exists already, it will generate a new one and check again."""
-            id = 'F' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(7))
-        else:
-            """since the id doesn't exist it returns it"""
-            return id
\ No newline at end of file
+		"""Checks if the random id created already exists in the array"""
+		while id in self.fellow_data:
```
Comment (mfalade) [inline models/fellow.py]: I'd suggest you you use `recursion` here instead.

You could decorate `get_fellow_id` as a class method, the have something like this

```python
@classmethod
def get_fellow_id(cls):
    fellow_id = 'F' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(7))
    if fellow_id in cls.fellow_data:
        cls.get_fellow_id()
    return fellow_id

```

File: models/fellow.py (hunk 17 above)
Comment (andela-bolajide) [inline models/fellow.py]: I've corrected that.

File: main.py (hunk 12 above)
Comment (andela-bolajide) [inline main.py]: Done. 
I have implemented the singleton pattern for instantiating my class.
Thanks.

File: models/fellow.py (hunk 18 above)
Comment (andela-bolajide) [inline models/fellow.py]: Done!

File: data/data_manager.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+
+    class __DatabaseManager:
+
+        def __init__(self):
+            self.name = 'Dojo'
+
+        def initialize(self):
+            # Create Office Table
+            self.cursor.execute(
+                '''CREATE TABLE IF NOT EXISTS office (room_name PRIMARY KEY,room_type,occupants,current_number,max_no_occupants)''')
```
Comment (tsolarin) [inline data/data_manager.py]: This SQL query is better as a variable, maybe a constant at the top?

File: data/data_manager.py [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+            DatabaseManager().close_conn()
+            print('Data has been successfully loaded into the app')
+
+    instance = None
+
+    def __init__(self):
+        if not DatabaseManager.instance:
+            DatabaseManager.instance = DatabaseManager.__DatabaseManager()
+        else:
+            pass
```
Comment (tsolarin) [inline data/data_manager.py]: No need for the `else` block here.

File: models/staff.py [hunk 21]
```diff
@@ -13,15 +14,21 @@ def __init__(self, first_name, last_name):
...
     def get_staff_id(self):
         """Returns unique id for room by generating random string"""
-        id = 'S' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(7))
+        staff_id = 'S' + \
+            ''.join(random.choice(string.ascii_uppercase + string.digits)
+                    for _ in range(7))
 
         """Checks if the random id created already exists in the array"""
-        while id in Staff.staff_data:
+        while staff_id in Staff.staff_data:
```
Comment (tsolarin) [inline models/staff.py]: Why not `self.staff_data`?

**Comment by tsolarin** [comment]
@andela-bolajide you should purge this repo of `.pyc` files

File: data/data_manager.py (hunk 20 above)
Comment (andela-bolajide) [inline data/data_manager.py]: I've removed it.

File: models/staff.py (hunk 21 above)
Comment (andela-bolajide) [inline models/staff.py]: Because i declared outside the __init__ statement but since it is also a class variable, i'll use self.staff_data.
Thanks

**Comment by njirap** [comment]
How to write git commits `https://chris.beams.io/posts/git-commit/`

File: .gitignore [hunk 22]
```diff
@@ -1,9 +1,15 @@
-virtual/
-data/
-dojo.db
 mock/
-workfile.txt
 __pycache__/
 models/__pycache__/
-tests/__pycache__/
 *.pyc
+models/*.pyc
```
Comment (njirap) [inline .gitignore]: isn't this already covered by L4

File: main.py [hunk 23]
```diff
@@ -65,33 +85,67 @@ def do_create_room(self, arg):
...
 
     @docopt_cmd
-    def do_add_person(self,arg):
+    def do_add_person(self, arg):
         """Usage: add_person <person_fname> <person_lname> <FELLOW/STAFF> [<wants_accommodation>]"""
         check = Dojo().add_person(arg)
-        if check == True:
-                Dojo().assign_room(arg)
+        if check:
+            Dojo().allocate_room(arg)
```
Comment (njirap) [inline main.py]: this looks funny..

File: main.py [hunk 24]
```diff
@@ -65,33 +85,67 @@ def do_create_room(self, arg):
...
     @docopt_cmd
-    def do_db_init(self,arg):
-        """Initialize the database for first time use. \
-        (This should only be done once to avoid complications."""
-        DatabaseInit().initialize()
+    def do_purge(self, arg):
+        """Usage: purge"""
+        print("Are you sure you want to purge the dojo?")
+        arg = input("Enter \'yes\' to approve and 'no' to reject: \n ")
+        Dojo().purge(arg)
```
Comment (njirap) [inline main.py]: maybe only purge when the `arg` is a `yes` instead of checking for the yes in the "backend"

File: models/fellow.py [hunk 25]
```diff
@@ -11,18 +12,22 @@ class Fellow(Person):
     def __init__(self, first_name, last_name):
         """Initializes a new fellow and assigns properties of the fellow"""
         super(Fellow, self).__init__(first_name, last_name)
-        self.person_type = 'fellow'
+        self.person_type = 'FELLOW'
         self.fellow_id = self.get_fellow_id()
         Fellow.fellow_data.append(self.fellow_id)
+        self.office_allocated = False
```
Comment (njirap) [inline models/fellow.py]: this can be gotten from `bool(self.office)`

File: models/fellow.py [hunk 26]
```diff
@@ -11,18 +12,22 @@ class Fellow(Person):
...
+    def get_fellow_id(cls):
+        # Returns unique id for Fellow by generating random string
+        fellow_id = 'F' + \
+            ''.join(random.choice(string.ascii_uppercase + string.digits)
+                    for _ in range(7))
+        # Checks if the random id created already exists in the array
+        if fellow_id in cls.fellow_data:
+            cls.get_fellow_id()
+        # It returns the id
+        return fellow_id
```
Comment (njirap) [inline models/fellow.py]: maybe append the new id to the `fellow_data` before you return it

File: models/fellow.py [hunk 27]
```diff
@@ -11,18 +12,22 @@ class Fellow(Person):
...
-            return id
\ No newline at end of file
+    @classmethod
+    def get_fellow_id(cls):
+        # Returns unique id for Fellow by generating random string
+        fellow_id = 'F' + \
+            ''.join(random.choice(string.ascii_uppercase + string.digits)
+                    for _ in range(7))
+        # Checks if the random id created already exists in the array
+        if fellow_id in cls.fellow_data:
```
Comment (njirap) [inline models/fellow.py]: maybe a better variable name than `fellow_data` to better represent what that variable is about..

File: models/livingspace.py [hunk 28]
```diff
@@ -1,29 +1,11 @@
...
-import string
-
 from models.room import Room
 
-class LivingSpace(Room):
-    """ This class creates a Living Space class which inherits from the Room Class"""
 
-    livingspace_data = []
+class LivingSpace(Room):
+    """This creates a LivingSpace class which inherits from the Room Class."""
```
Comment (njirap) [inline models/livingspace.py]: """
Documentation goes here
"""

File: models/staff.py [hunk 29]
```diff
@@ -2,26 +2,33 @@
...
+                    for _ in range(7))
 
         """Checks if the random id created already exists in the array"""
-        while id in Staff.staff_data:
-            """if the array exists already, it will generate a new one and check again."""
-            id = 'S' + ''.join(random.choice(string.ascii_uppercase + string.digits) for _ in range(7))
+        while staff_id in self.staff_data:
+            """if the id exists, it will generate a new one and check again."""
+            staff_id = 'S' + \
+                ''.join(random.choice(string.ascii_uppercase + string.digits)
```
Comment (njirap) [inline models/staff.py]: Practice being `DRY` in your code

File: test_input3.txt [hunk 30]
```diff
@@ -0,0 +1,2 @@
+BOLAJI OLAJIDE FELLOW
+BOLAJI OLAJIDE STAFF
```
Comment (njirap) [inline test_input3.txt]: Maybe we just need 1 test_input files..

File: tests/test_fellow.py [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+class TestFellow(unittest.TestCase):
+    """Test cases for the Fellow class"""
+
+    def setUp(self):
+        """Setup a new Fellow object to be used for testing"""
+        self.fellow = Fellow('Brian', 'Mosigisi')
+
+    def test_fellow_type(self):
+        """Tests if the object created is of type Fellow"""
+        self.assertEqual(type(self.fellow), Fellow)
```
Comment (njirap) [inline tests/test_fellow.py]: under what circumstances would you instantiate `Fellow` and that instance not be an instance of `Fellow`?

File: .gitignore (hunk 22 above)
Comment (andela-bolajide) [inline .gitignore]: I have removed all duplicates.

File: main.py (hunk 23 above)
Comment (andela-bolajide) [inline main.py]: Hello @njirap,
When i changed to `Dojo.allocate_room(arg)` i get an error:

screen shot 2017-03-08 at 5 59 26 pm

Is it better if i probably allocate the instance of Dojo() to a variable? Something like:

`
self.dojo = Dojo()
self.dojo.allocate_room(arg)
`

File: main.py (hunk 24 above)
Comment (andela-bolajide) [inline main.py]: Done.

File: models/fellow.py (hunk 25 above)
Comment (andela-bolajide) [inline models/fellow.py]: Done.

File: models/fellow.py (hunk 26 above)
Comment (andela-bolajide) [inline models/fellow.py]: Done.

File: models/fellow.py (hunk 27 above)
Comment (andela-bolajide) [inline models/fellow.py]: Done.

File: models/staff.py (hunk 29 above)
Comment (andela-bolajide) [inline models/staff.py]: Done.

File: tests/test_fellow.py (hunk 31 above)
Comment (andela-bolajide) [inline tests/test_fellow.py]: I have tested if the Fellow class is inheriting from the Person class.
